hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -2312,17 +2311,40 @@ private void listStatusFindKeyInTableCache(
     return fileStatusList;
   }
 
+  @SuppressWarnings("methodlength")
   public List<OzoneFileStatus> listStatusV1(OmKeyArgs args, boolean recursive,
       String startKey, long numEntries, String clientAddress)
           throws IOException {
     Preconditions.checkNotNull(args, "Key args can not be null");
 
     // unsorted OMKeyInfo list contains combine results from TableCache and DB.
     List<OzoneFileStatus> fileStatusFinalList = new ArrayList<>();
-    LinkedHashSet<OzoneFileStatus> fileStatusList = new LinkedHashSet<>();
+
     if (numEntries <= 0) {
       return fileStatusFinalList;
     }
+
+    /**
+     * A map sorted by OmKey to combine results from TableCache and DB for
+     * each entity - Dir & File.
+     *
+     * Two separate maps are required because the order of seek -> (1)Seek
+     * files in fileTable (2)Seek dirs in dirTable.
+     *
+     * StartKey should be added to the final listStatuses, so if we combine
+     * files and dirs into a single map then directory with lower precedence
+     * will appear at the top of the list even if the startKey is given as
+     * fileName.
+     *
+     * For example, startKey="a/file1". As per the seek order, first fetches
+     * all the files and then it will start seeking all the directories.
+     * Assume a directory name exists "a/b". With one map, the sorted list will
+     * be ["a/b", "a/file1"]. But the expected list is: ["a/file1", "a/b"],
+     * startKey element should always be at the top of the listStatuses.
```

Comment:
A little confused with this example. startKey is used to list keys which are behinds this key. "a/b" should not be expected in the return list, right? I also checked the logic in getDirectories method. It also doesn't use startKey to do the key comparison, [KeyManagerImpl.java#L2633](https://github.com/apache/ozone/blob/HDDS-2939/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2633)

Reply:
Earlier the files and dirs were kept in single keyTable it was easy for the traversal and easy to do lexicographically greater than the startKey argument and RocksDB filters were helping us to maintain sorted order. 

But in new code, since we have two different tables(file & dir) we are iterating separately and programatically I've a search algorithm were it first fetches entries from the file and then fetches from dir Table. So, the combined list won't be in sorted order.
In V1 : the final list will be like [all files, all sub-dirs]. ["a/file1", "a/b"]. 

[Reference](https://github.com/apache/ozone/blob/HDDS-2939/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2424)

I understand, its hard to understand. I think, we can ensure correctness in first phase. Later, in folow-up tasks we need to re-visit this part and try to optimize it further for better performance or so.

Doe this make sense to you?

Thanks again for the detailed reviews.

Reply:
For example, there are keys in FS:
a/b/c/file-1

Reply:
Thanks for detailed explanation, @rakeshadr . Suppose there is a different meaning of the startKey under separate  file/dir table here. I will take a deep look for this.
In order not to unblock subsequent PRs, I will give my approval here.

Reply:
@linyiqun Thanks a lot. Sure, will continue the discussions and refine it, if there are any gaps in the `startKey` semantics.